lib/src/container/import.rs

```diff
@@ -39,31 +41,120 @@ async fn fetch_manifest(imgref: &ImageReference) -> Result<(oci::Manifest, Strin
+    let (pipein, input_copydriver) = crate::async_util::copy_async_read_to_sync_pipe(src)?;
+    let (tx_buf, rx_buf) = tokio::sync::mpsc::channel(2);
+    let blob_symlink_target = format!("../{}.tar", blobid);
+    let import = tokio::task::spawn_blocking(move || {
+        let mut archive = tar::Archive::new(pipein);
+        let mut buf = vec![0u8; 8192];
+        for entry in archive.entries()? {
+            let mut entry = entry.context("Reading entry")?;
+            let path = entry.path()?;
+            let path = &*path;
+            let path = Utf8Path::from_path(path)
+                .ok_or_else(|| anyhow!("Invalid non-utf8 path {:?}", path))?;
+            let t = entry.header().entry_type();
+
+            // We generally expect our layer to be first, but let's just skip anything
+            // unexpected to be robust against changes in skopeo.
+            if path.extension() != Some("tar") {
+                continue;
+            }
+
+            match t {
+                tar::EntryType::Symlink => {
+                    if let Some(name) = path.file_name() {
+                        if name == "layer.tar" {
+                            let target = entry
+                                .link_name()?
+                                .ok_or_else(|| anyhow!("Invalid link {}", path))?;
+                            let target = Utf8Path::from_path(&*target)
+                                .ok_or_else(|| anyhow!("Invalid non-UTF8 path {:?}", target))?;
+                            if target != blob_symlink_target {
+                                return Err(anyhow!(
+                                    "Found unexpected layer link {} -> {}",
+                                    path,
+                                    target
+                                ));
+                            }
+                        }
+                    }
+                }
+                tar::EntryType::Regular => loop {
+                    let n = entry
+                        .read(&mut buf[..])
+                        .context("Reading tar file contents")?;
+                    let done = 0 == n;
+                    let r = Ok::<_, std::io::Error>(bytes::Bytes::copy_from_slice(&buf[0..n]));
+                    let receiver_closed = tx_buf.blocking_send(r).is_err();
+                    if receiver_closed || done {
+                        return Ok::<_, anyhow::Error>(());
+                    }
+                },
+                _ => continue,
+            }
+        }
+        Err(anyhow!("Failed to find layer {}", blob_symlink_target))
+    })
+    .map_err(anyhow::Error::msg);
+    let stream = tokio_stream::wrappers::ReceiverStream::new(rx_buf);
+    let reader = tokio_util::io::StreamReader::new(stream);
+    // Is there a better way to do this?
+    let worker = async move {
+        let (import, input_copydriver) = tokio::join!(import, input_copydriver);
+        let _: () = import?.context("Import worker")?;
```

Comment:
Just wondering what the purpose of these unused variables are. Is it just for adding annotations for the compiler?

Reply:
It's confusing because there's `Result<Result<T>>` stuff involved here and I want to be careful we're fully unwrapping these and handling the chain.

It's too easy to do e.g. `let _ = foo?` when it should be `let _ = foo??`.

Reply:
To elaborate further by doing `let _: ()` and adding the explicit `()` typing we make it a compile error if the result of `?` isn't `()`.

Reply:
makes sense, thanks!